metaflow/plugins/project_decorator.py

```diff
@@ -0,0 +1,98 @@
+                 "not specified.")
+    }
+
+    def flow_init(self,
+                  flow,
+                  graph,
+                  environment,
+                  datastore,
+                  logger,
+                  echo,
+                  options):
+        self._option_values = options
+        project_name = self.attributes.get('name')
+        project_flow_name, branch_name = format_name(flow.name,
+                                                     project_name,
+                                                     options['production'],
+                                                     options['branch'],
+                                                     get_username())
+        is_user_branch = options['branch'] is None and not options['production']
+        current._update_env({'project_name': project_name,
+                             'branch_name': branch_name,
+                             'is_user_branch': is_user_branch,
+                             'is_production': options['production'],
+                             'project_flow_name': project_flow_name})
+
+    def get_top_level_options(self):
+        return list(self._option_values.items())
+
+def format_name(flow_name,
+                project_name,
+                deploy_prod,
+                given_branch,
+                user_name):
+
+    if not project_name:
+        # an empty string is not a valid project name
+        raise MetaflowException("@project needs a name. "
+                                "Try @project(name='some_name')")
+    elif re.search(VALID_NAME_RE, project_name):
+        raise MetaflowException("The @project name must contain only "
+                                "lowercase alphanumeric characters "
+                                "and underscores.")
+    elif len(project_name) > VALID_NAME_LEN:
+        raise MetaflowException("The @project name must be shorter than "
+                                "%d characters." % VALID_NAME_LEN)
+
+    if given_branch and deploy_prod:
+        raise MetaflowException("Specify either --production or --branch.")
+
+    if given_branch:
+        if re.search(VALID_NAME_RE, given_branch):
+            raise MetaflowException("The branch name must contain only "
+                                    "lowercase alphanumeric characters "
+                                    "and underscores.")
+        elif len(given_branch) > VALID_NAME_LEN:
+            raise MetaflowException("Branch name is too long. "
+                                    "The maximum is %d characters."\
+                                    % VALID_NAME_LEN)
+
+        branch = 'test.%s' % given_branch
+    elif deploy_prod:
+        branch = 'prod'
+    else:
+        # For AWS Step Functions, we set the branch to the value of
+        # environment variable `METAFLOW_OWNER`, since AWS Step Functions
+        # has no notion of user name.
+        branch = 'user.%s' % os.environ.get('METAFLOW_OWNER', user_name)
```

Comment:
I assume `METAFLOW_OWNER` may have dots in it, is it going to cause issues? If not should we allow dots in branch names?

Reply:
Good point! There is actually an open issue to sanitize `METAFLOW_OWNER` since it can have whitespace as well. I will open a different PR to make sure we are sanitizing the output of `utils.get_username()`

Reply:
Maybe validate it against `VALID_NAME_RE` here then? I worry that we'll ship this PR, then people will use it to create branches with weird names, then we ship sanitization and it will be annoying to reconcile their existing branch names with the sanitized ones.

Reply:
Yeah, the check needs to exist alongside the `get_username()` method. We will ship that PR and this PR in a single release.

Reply:
@oavdeev Just so that we are on the same page - any user name that contributes to a valid step functions state machine name is allowed - which means whitespace, wildcard characters, brackets and special characters are not allowed (and currently the PR will throw a not so pretty error) - dots, hyphens and dashes are still allowed and will work as is. So I don't see the risk of shipping this PR as is.

Reply:
There is a similar issue that has existed with @batch for a while now where we are unable to launch the AWS Batch job if there are white spaces in the user name.

Reply:
Ah I thought you'd replace the dots in the username. I.e. if the user is `oleg.avdeev` the branch would be `oleg_avdeev` or something.